### PR TITLE
Provide support for Sticky ad units in AMP #2472

### PR DESCRIPTION
**Pull request by zhouyx**
Initial implementation of #2472

### CODE REVIEW

**Comment by zhouyx** [comment]
@cramforce @erwinmombay PTAL
Add the extension file. Call the new extension amp-stick-ad now. Should I rename it to something like amp-stick-ad-container? since it's actually a container for ad.

File: extensions/amp-sticky-ad/0.1/amp-sticky-ad.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+/** @const */
+const EXPERIMENT = 'amp-sticky-ad';
+/** @const */
+const TAG = 'amp-sticky-ad';
+
+class AmpStickyAd extends AMP.BaseElement {
+  /** @override */
+  isLayoutSupported() {
+    return true;
```
Comment (erwinmombay) [inline extensions/amp-sticky-ad/0.1/amp-sticky-ad.js]: should this just be `layout == Layout.FIXED;`

File: tools/experiments/experiments.js [hunk 2]
```diff
@@ -73,6 +73,11 @@ const EXPERIMENTS = [
     name: 'Viewability APIs for amp-analytics',
     spec: 'https://github.com/ampproject/amphtml/issues/1297#issuecomment-197441289',
   },
+  {
+    id: 'amp-sticky-ad',
+    name: 'AMP Sticky Ad',
+    spec: '',
```
Comment (erwinmombay) [inline tools/experiments/experiments.js]: is there an intent to implement or design doc anywhere?

File: examples/sticky-ads-amp.html [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      margin: 18px 0;
+      padding: 0;
+    }
+
+
+
+  </style>
+  <script async custom-element="amp-sticky-ad" src="https://cdn.ampproject.org/v0/amp-sticky-ad-0.1.js"></script>
+  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+
```
Comment (erwinmombay) [inline examples/sticky-ads-amp.html]: remove erroneous spaces

File: examples/sticky-ads-amp.html [hunk 4]
```diff
@@ -0,0 +1,52 @@
+<!doctype html>
+<html ⚡>
+<head>
+  <meta charset="utf-8">
+  <title>Ad examples</title>
+  <link rel="canonical" href="http://nonblocking.io/" >
+  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+  <style amp-custom>
+    amp-sticky-ad {
```
Comment (erwinmombay) [inline examples/sticky-ads-amp.html]: will this actually be the css for the element? if so, this should cs inside `amp-stick-ad.css`

File: extensions/amp-sticky-ad/0.1/amp-sticky-ad.js (hunk 1 above)
Comment (erwinmombay) [inline extensions/amp-sticky-ad/0.1/amp-sticky-ad.js]: should this just be `layout == Layout.FIXEDHEIGHT`?

File: tools/experiments/experiments.js (hunk 2 above)
Comment (cramforce) [inline tools/experiments/experiments.js]: There is: https://github.com/ampproject/amphtml/issues/2472

Speaking of which:
@zhouyx: Please add something like this to the description: "Initial implementation of #2472"

This links the pull request to the issue (In both directions) which can be very helpful for people to follow the work.

File: extensions/amp-sticky-ad/0.1/amp-sticky-ad.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {isExperimentOn} from '../../../src/experiments';
+import {dev} from '../../../src/log';
+import {Layout} from '../../../src/layout';
+import {CSS} from '../../../build/amp-sticky-ad-0.1.css';
+
+/** @const */
+const EXPERIMENT = 'amp-sticky-ad';
+/** @const */
+const TAG = 'amp-sticky-ad';
```
Comment (erwinmombay) [inline extensions/amp-sticky-ad/0.1/amp-sticky-ad.js]: lets just remove experiment and use TAG as well for the experiment name. these two used to be different but this was actually your first task to rename camelCase to dash-case

https://github.com/ampproject/amphtml/pull/2837/files

File: tools/experiments/experiments.js [hunk 6]
```diff
@@ -73,6 +73,11 @@ const EXPERIMENTS = [
     name: 'Viewability APIs for amp-analytics',
     spec: 'https://github.com/ampproject/amphtml/issues/1297#issuecomment-197441289',
   },
+  {
+    id: 'amp-sticky-ad',
+    name: 'AMP Sticky Ad',
+    spec: 'https://github.com/ampproject/amphtml/issues/2472',
```
Comment (zhouyx) [inline tools/experiments/experiments.js]: Do I need a include a .md file here? Or the link to the issue is good?

**Comment by zhouyx** [comment]
@sriramkrish85 PTAL Thanks!

File: tools/experiments/experiments.js (hunk 6 above)
Comment (erwinmombay) [inline tools/experiments/experiments.js]: link to issue should be good enough

Comment (sriramkrish85) [inline tools/experiments/experiments.js]: Just add an empty .md file and link it here.

File: examples/sticky.ads.amp.html [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  <title>Ad examples</title>
+  <link rel="canonical" href="http://nonblocking.io/" >
+  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+  <script async custom-element="amp-sticky-ad" src="https://cdn.ampproject.org/v0/amp-sticky-ad-0.1.js"></script>
+  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+
+</head>
+<body>
+  <h2>Adman sticky</h2>
```
Comment (sriramkrish85) [inline examples/sticky.ads.amp.html]: Sticky Ad Demo

File: extensions/amp-sticky-ad/0.1/amp-sticky-ad.css [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ amp-sticky-ad {
+ 	width: 100%;
```
Comment (sriramkrish85) [inline extensions/amp-sticky-ad/0.1/amp-sticky-ad.css]: why?

Comment (zhouyx) [inline extensions/amp-sticky-ad/0.1/amp-sticky-ad.css]: I thought we want the sticky ad width to always be 100%

File: tools/experiments/experiments.js (hunk 6 above)
Comment (zhouyx) [inline tools/experiments/experiments.js]: Added an empty .md file. But I feel the content in the link explain `amp-sticky-ad` better so I leave the link here.

File: extensions/amp-sticky-ad/0.1/amp-sticky-ad.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import {CSS} from '../../../build/amp-sticky-ad-0.1.css';
+import {Layout} from '../../../src/layout';
+import {dev} from '../../../src/log';
+import {isExperimentOn} from '../../../src/experiments';
+import {isLayoutSizeDefined} from '../../../src/layout';
+
+/** @const */
+const TAG = 'amp-sticky-ad';
+
+class AmpStickyAd extends AMP.BaseElement {
```
Comment (erwinmombay) [inline extensions/amp-sticky-ad/0.1/amp-sticky-ad.js]: add space

**Pull request merged**